Add rich logging configuration (#31205)
ray-project/ray: doc/source/ray-observability/ray-logging.rst

=======
This document will explain Ray's logging system and its best practices.

Internal Ray Logging Configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
When ``import ray`` is executed, Ray's logging system is initialized, generating a sensible configuration given in ``python/ray/log.py``. The default logging level is ``logging.INFO``.

> 

Reviewer: Can you add examples here?

Ray uses `rich <https://github.com/Textualize/rich>`__ to generate log messages that are well formatted and easily identified. If ``rich`` is not installed, a simple fallback formatter is used. Rich log messages display contextual information including the Ray component from which the message was generated and the filename and line number.

> Changing Ray's Internal Logging Configuration

Reviewer: Can we just remove the whole section and just add how to configure library level logging through a  code example from the above section? I think everything here is unnecessary details for users. We need only these 3 information. 

1. All "ray" loggers are automatically configured and the config is specified in log.py
2. You can change the ray library logging config following the code below.
```
logger = logging.getLogger("ray")
logger # Modify the logging config
```
3. Add examples for other ray libraries (tune, data, rllib, train, and workflow). 
4. If you'd like to propagate the config to workers, follow `Worker logs` section

> Ray uses `rich <https://github.com/Textualize/rich>`__ to generate log messages that are well formatted and easily identified. If ``rich`` is not installed, a simple fallback formatter is used. Rich log messages display contextual information including the Ray component from which the message was generated as well as the filename and line number which are not available with the simple fallback logger.

Reviewer 2: ```suggestion
Ray uses `rich <https://github.com/Textualize/rich>`__ to generate log messages that are well formatted and easily identified. If ``rich`` is not installed, a simple fallback formatter is used. Rich log messages display contextual information including the Ray component from which the message was generated and the filename and line number.
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Modifying Ray's logging configuration can be done by changing the configuration in ``python/ray/log.py`` according to `the python logging documentation <https://docs.python.org/3/howto/logging.html>`__. Ray uses a `dictConfig <https://docs.python.org/3/library/logging.config.html#logging.config.dictConfig>`__ to specify logging configuration.

Reviewer 2: ```suggestion
To modify Ray's logging configuration, change the configuration in ``python/ray/log.py`` according to `the python logging documentation <https://docs.python.org/3/howto/logging.html>`__. Ray uses a `dictConfig <https://docs.python.org/3/library/logging.config.html#logging.config.dictConfig>`__ to specify logging configuration.
```

> Logging configuration should be kept inside ``python/ray/log.py`` to avoid spreading conflicting logging configurations across the database. There is no need to call ``logging.basicConfig`` because logging configuration is done at import time.

Reviewer: ```suggestion
Logging configuration should be kept inside ``python/ray/log.py``. 
```

> Logging configuration should be kept inside ``python/ray/log.py`` to avoid spreading conflicting logging configurations across the database. Generally speaking there is no need to call ``logging.basicConfig``, as logging configuration is done at import time.

Reviewer 2: ```suggestion
Logging configuration should be kept inside ``python/ray/log.py`` to avoid spreading conflicting logging configurations across the database. You do not need to call ``logging.basicConfig`` because logging configuration is done at import time.
```

.. note::

>    The one exception to this is if logging configuration is needed in a separate process. Since the Python logger module creates a singleton logger per process, if a worker needs to log a message they should configure logging on a per task/actor basis. See information further down in the :ref:`Customizing worker loggers <customizing-worker-loggers> section.`

Reviewer 2: ```suggestion
   The exception is if you need logging configuration in a separate process. Since the Python logger module creates a singleton logger per process, if a worker needs to log a message it should configure logging on a per task or actor basis. For more information, see :ref:`Customizing worker loggers <customizing-worker-loggers> section.`
```

> Additionally, do not log Ray log messages to the root logger. `Always to a logger with the current file name <https://docs.python.org/3/howto/logging.html#advanced-logging-tutorial>`__:

Reviewer: This doesn't seem to be necessary for users to know. Can we remove  this section?

> Additionally, internally Ray log messages should not be logged to the root logger - `always to a logger with the current file name <https://docs.python.org/3/howto/logging.html#advanced-logging-tutorial>`__:

Reviewer 2: ```suggestion
Additionally, do not log Ray log messages to the root logger. `Always log to a logger with the current file name <https://docs.python.org/3/howto/logging.html#advanced-logging-tutorial>`__:
```

.. _ray-worker-logs:

Worker stdout and stderr
~~~~~~~~~~~
Ray's tasks or actors are executed remotely within Ray's worker processes. Ray has special support to improve the visibility of stdout and stderr produced by workers.

> - By default, all of the tasks/actors stdout and stderr are redirected to the worker log files. This includes any log messages generated by the worker. Check out :ref:`Logging directory structure <logging-directory-structure>` to learn how Ray's logging directory is structured.

Reviewer 2: ```suggestion
- By default, stdout and stderr from all tasks and actors are redirected to the worker log files, including any log messages generated by the worker. See :ref:`Logging directory structure <logging-directory-structure>` to understand the structure of the Ray logging directory.
```

> - By default, all of the tasks/actors stdout and stderr that is redirected to worker log files is read by the driver and displayed with a call to ``print``. Drivers display all stdout/stderr generated from its tasks/actors to its own stdout/stderr.

Reviewer 2: ```suggestion
- By default, the driver reads the worker log files to which the stdout and stderr for all tasks and actors are redirected. To display the log files, call `print`. Drivers display all stdout and stderr generated from their tasks and actors to their own stdout and stderr.
```

Author: | To display the log files, call `print`

Ahh, this isn't what I meant by this passage. I meant that internally, Ray calls `print` to display stdout and stderr which are stream to the log files from the worker.

IMO this is something of an anti-pattern because it bypasses python's logging system and introduces a number of difficulties in configuring loggers. The reason I mention it here is that the fact that log file contents are printed to the stdout of the driver has implications for the user, and helps to understand the current logging system.

Reviewer 2: Ah! I mangled that good!. I think you worded it really well in the comment, `Ray calls print to display stdout and stderr which are streamed to the log files from the worker.`

> - By default, the driver reads the worker log files to which the stdout and stderr for all tasks and actors are redirected. Drivers display all stdout and stderr generated from their tasks or actors to their own stdout and stderr by passing log file content to ``print``.

Reviewer: ```suggestion
- By default, the driver reads the worker log files to which the stdout and stderr for all tasks and actors are redirected. Drivers display all stdout and stderr generated from their tasks or actors to their own stdout and stderr
```

This is an impl detail

    (MyActor(index=2) pid=482120) hello there
    (MyActor(index=1) pid=482119) hello there

Customizing worker loggers
~~~~~~~~~~~~~~~~~~~~~

> When using ray, all of the tasks and actors are executed remotely in Ray's worker processes.

Reviewer 2: ```suggestion
When using Ray, all tasks and actors are executed remotely in Ray's worker processes.
```

> If a user intends to log a message via a worker, a logger should be used. Typically the way this is done is by getting a logger using the filename where the log is generated; see `the Python logging documentation <https://docs.python.org/3/howto/logging.html#advanced-logging-tutorial>`__ for more information about this.

Reviewer 2: ```suggestion
If you intend to log a message via a worker, use a logger. Get a logger using the filename of the generated log; see `the Python logging documentation <https://docs.python.org/3/howto/logging.html#advanced-logging-tutorial>`__ for more information.
```

Author: Hmm, I've reworded slightly to avoid using 'log' too much here. Let me know if you'd prefer something different.

Reviewer 2: Much better!

> By default, workers use the standard Ray logging configuration given in `python/ray/log.py` for code that is internal to Ray. Because the Python logger module creates a singleton logger per process, if a worker needs to log a message it should configure logging on a per task or actor basis.

Reviewer: I don't think this section is necessary (the section above is sufficient). Or just have a single sentence "to change ray library logging config, see X"

> Workers will by default use the standard Ray logging configuration given in `python/ray/log.py` for code that is internal to Ray. Since the Python logger module creates a singleton logger per process, if a worker needs to log a message they should configure logging on a per task/actor basis.

Reviewer 2: ```suggestion
By default, workers use the standard Ray logging configuration given in `python/ray/log.py` for code that is internal to Ray. Because the Python logger module creates a singleton logger per process, if a worker needs to log a message they should configure logging on a per task or actor basis.
```

.. note::

>    Loggers configured this way are not child loggers of Ray's internal logging system. They will not use the internal Ray formatter.

Reviewer: ```suggestion

```

>    Loggers configured this way are not child loggers of Ray's internal logging system. As such, they will not use the internal Ray formatter.

Reviewer 2: ```suggestion
   Loggers configured this way are not child loggers of Ray's internal logging system. They will not use the internal Ray formatter.
```

            logging.basicConfig(level=logging.INFO)

        def log(self, msg):
            logger = logging.getLogger(__name__)
            logger.info(msg)

    actor = Actor.remote()
    ray.get(actor.log.remote("A log message for an actor."))

    @ray.remote
    def f(msg):
        logging.basicConfig(level=logging.INFO)
        logger = logging.getLogger(__name__)
        logger.info(msg)

    ray.get(f.remote("A log message for a task."))

.. code-block:: bash

    (Actor pid=179641) INFO:__main__:A log message for an actor.

>     (f pid=177572) INFO:__main__:A log message for a task.

Reviewer: Hmm we shouldn't touch the user level logging at all? Let's discuss this part in person

Author: We discussed - we're going to keep this, as it follows python logging documentation.
